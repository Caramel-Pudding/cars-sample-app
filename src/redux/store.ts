import { configureStore, ThunkAction, Action } from "@reduxjs/toolkit";
import { reducer as filtersReducer } from "./features/filters/slice";
import { reducer as paginationReducer } from "./features/pagination/slice";
import { reducer as sortingReducer } from "./features/sorting/slice";
import { reducer as carsReducer } from "./features/cars/slice";

export const store = configureStore({
  reducer: {
    filters: filtersReducer,
    pagination: paginationReducer,
    sorting: sortingReducer,
    cars: carsReducer,
  },
});

export type AppDispatch = typeof store.dispatch;
export type RootState = Readonly<ReturnType<typeof store.getState>>;
export type AppThunk<ReturnType = void> = ThunkAction<
  ReturnType,
  RootState,
  unknown,
  Action<string>
>;
