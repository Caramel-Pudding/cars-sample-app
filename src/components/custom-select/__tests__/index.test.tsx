import React, { FC, useState } from "react";
import { render, cleanup, fireEvent } from "@testing-library/react";
import { MemoryRouter as Router } from "react-router-dom";
import { Provider } from "react-redux";

import { store } from "../../../redux/store";
import { CustomSelect, CustomSelectProps, inputDataTestId } from "..";

describe("<CustomSelect />", () => {
  afterEach(cleanup);

  const options = ["TEST1", "TEST2"];
  const testLabel = "TEST";
  const placeholder = "placeholder";

  const TestWrapper: FC<Partial<CustomSelectProps>> = () => {
    const [selectValue, setSelectValue] = useState(options[0]);

    return (
      <Provider store={store}>
        <Router>
          <CustomSelect
            placeholder={placeholder}
            chosenOption={selectValue}
            handler={setSelectValue}
            labelText={testLabel}
            options={options}
          />
        </Router>
      </Provider>
    );
  };

  it("should render proper label", () => {
    // * #TEST: ARRANGE
    const { container, getByText } = render(<TestWrapper />, {});
    const label = getByText(testLabel);
    // * #TEST: ASSERT
    expect(label).toHaveTextContent(testLabel);
    expect(container).toMatchSnapshot();
  });

  it("should render undefined label as ''", () => {
    // * #TEST: ARRANGE
    const strange = render(<TestWrapper labelText={undefined} />, {});
    const common = render(<TestWrapper labelText="" />, {});

    // * #TEST: ASSERT
    expect(strange.container).toEqual(common.container);
    expect(strange.container).toMatchSnapshot();
    expect(common.container).toMatchSnapshot();
  });

  it("Properly select another option", () => {
    // * #TEST: ARRANGE
    const { container, getByText, getByTestId } = render(<TestWrapper />);

    // * #TEST: ASSERT
    expect(getByTestId(inputDataTestId)).toHaveTextContent(options[0]);
    expect(container).toMatchSnapshot();

    // * #TEST: ACT

    fireEvent.click(getByTestId(inputDataTestId));
    fireEvent.click(getByText(options[1])!);

    // * #TEST: ASSERT
    expect(getByTestId(inputDataTestId)).toHaveTextContent(options[1]);
    expect(container).toMatchSnapshot();
  });

  it("Uncheck when already selected option clicked", () => {
    // * #TEST: ARRANGE
    const { container, getAllByText, getByTestId } = render(<TestWrapper />);

    // * #TEST: ASSERT
    expect(getByTestId(inputDataTestId)).toHaveTextContent(options[0]);
    expect(container).toMatchSnapshot();

    // * #TEST: ACT

    fireEvent.click(getByTestId(inputDataTestId));
    fireEvent.click(getAllByText(options[0])[1]);

    // * #TEST: ASSERT
    expect(getByTestId(inputDataTestId)).toHaveTextContent(placeholder);
    expect(container).toMatchSnapshot();
  });
});
