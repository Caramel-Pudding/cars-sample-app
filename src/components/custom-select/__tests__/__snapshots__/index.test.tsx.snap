// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CustomSelect /> Properly select another option 1`] = `
<div>
  <label
    class="container"
  >
    TEST
    <button
      class="input robotoMd"
      data-testid="INPUT_DATA_TEST_ID"
      tabindex="0"
      type="button"
    >
      TEST1
      <svg
        class="bullet"
        height="15"
        tabindex="0"
        width="15"
      >
        <polygon
          points="1,1 7,10 14,1"
          style="fill: #EDEDED;"
        />
      </svg>
    </button>
  </label>
</div>
`;

exports[`<CustomSelect /> Properly select another option 2`] = `
<div>
  <label
    class="container"
  >
    TEST
    <button
      class="input robotoMd"
      data-testid="INPUT_DATA_TEST_ID"
      tabindex="0"
      type="button"
    >
      TEST2
      <svg
        class="bullet bulletActive"
        height="15"
        tabindex="0"
        width="15"
      >
        <polygon
          points="1,1 7,10 14,1"
          style="fill: #EDEDED;"
        />
      </svg>
    </button>
    <ul
      class="list"
    >
      <li
        class="listElement listElementNotSelected"
        value="TEST1"
      >
        <button
          class="robotoMd button"
          type="button"
        >
          TEST1
        </button>
      </li>
      <li
        class="listElement listElementSelected"
        value="TEST2"
      >
        <button
          class="robotoMd button"
          type="button"
        >
          TEST2
        </button>
      </li>
    </ul>
  </label>
</div>
`;

exports[`<CustomSelect /> Uncheck when already selected option clicked 1`] = `
<div>
  <label
    class="container"
  >
    TEST
    <button
      class="input robotoMd"
      data-testid="INPUT_DATA_TEST_ID"
      tabindex="0"
      type="button"
    >
      TEST1
      <svg
        class="bullet"
        height="15"
        tabindex="0"
        width="15"
      >
        <polygon
          points="1,1 7,10 14,1"
          style="fill: #EDEDED;"
        />
      </svg>
    </button>
  </label>
</div>
`;

exports[`<CustomSelect /> Uncheck when already selected option clicked 2`] = `
<div>
  <label
    class="container"
  >
    TEST
    <button
      class="input robotoMd"
      data-testid="INPUT_DATA_TEST_ID"
      tabindex="0"
      type="button"
    >
      placeholder
      <svg
        class="bullet bulletActive"
        height="15"
        tabindex="0"
        width="15"
      >
        <polygon
          points="1,1 7,10 14,1"
          style="fill: #EDEDED;"
        />
      </svg>
    </button>
    <ul
      class="list"
    >
      <li
        class="listElement listElementNotSelected"
        value="TEST1"
      >
        <button
          class="robotoMd button"
          type="button"
        >
          TEST1
        </button>
      </li>
      <li
        class="listElement listElementNotSelected"
        value="TEST2"
      >
        <button
          class="robotoMd button"
          type="button"
        >
          TEST2
        </button>
      </li>
    </ul>
  </label>
</div>
`;

exports[`<CustomSelect /> should render proper label 1`] = `
<div>
  <label
    class="container"
  >
    TEST
    <button
      class="input robotoMd"
      data-testid="INPUT_DATA_TEST_ID"
      tabindex="0"
      type="button"
    >
      TEST1
      <svg
        class="bullet"
        height="15"
        tabindex="0"
        width="15"
      >
        <polygon
          points="1,1 7,10 14,1"
          style="fill: #EDEDED;"
        />
      </svg>
    </button>
  </label>
</div>
`;

exports[`<CustomSelect /> should render undefined label as '' 1`] = `
<div>
  <label
    class="container"
  >
    TEST
    <button
      class="input robotoMd"
      data-testid="INPUT_DATA_TEST_ID"
      tabindex="0"
      type="button"
    >
      TEST1
      <svg
        class="bullet"
        height="15"
        tabindex="0"
        width="15"
      >
        <polygon
          points="1,1 7,10 14,1"
          style="fill: #EDEDED;"
        />
      </svg>
    </button>
  </label>
</div>
`;

exports[`<CustomSelect /> should render undefined label as '' 2`] = `
<div>
  <label
    class="container"
  >
    TEST
    <button
      class="input robotoMd"
      data-testid="INPUT_DATA_TEST_ID"
      tabindex="0"
      type="button"
    >
      TEST1
      <svg
        class="bullet"
        height="15"
        tabindex="0"
        width="15"
      >
        <polygon
          points="1,1 7,10 14,1"
          style="fill: #EDEDED;"
        />
      </svg>
    </button>
  </label>
</div>
`;
